server/src/main/java/org/opensearch/rest/NamedRoute.java

```diff
@@ -9,47 +9,168 @@
 package org.opensearch.rest;
 
 import org.opensearch.OpenSearchException;
+import org.opensearch.transport.TransportService;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.function.Function;
 
 /**
  * A named Route
  *
  * @opensearch.internal
  */
 public class NamedRoute extends RestHandler.Route {
+
     private static final String VALID_ACTION_NAME_PATTERN = "^[a-zA-Z0-9:/*_]*$";
     static final int MAX_LENGTH_OF_ACTION_NAME = 250;
 
-    private final String name;
+    private final String uniqueName;
+    private final Set<String> actionNames;
 
-    public boolean isValidRouteName(String routeName) {
-        if (routeName == null || routeName.isBlank() || routeName.length() > MAX_LENGTH_OF_ACTION_NAME) {
-            return false;
+    private Function<RestRequest, RestResponse> handler;
+
+    /**
+     * Builder class for constructing instances of {@link NamedRoute}.
+     */
+    public static class Builder {
+        private RestRequest.Method method;
+        private String path;
+        private String uniqueName;
+        private final Set<String> legacyActionNames = new HashSet<>();
+        private Function<RestRequest, RestResponse> handler;
```

Comment:
Note to self: In Extensions, we will want to use an `ExtensionRestResponse` here (in order to track consumed content and params) but this syntax doesn't require it.  Since `NamedRoute` won't allow subclassing, we'll need to add some explicit type checking and documentation to make sure handlers conform to the required syntax.  

Might be worth an issue on SDK to refactor to check consumed params there and just use a regular `RestResponse`.